### ISSUE TITLE
Circe derivation fails on traits having children with same name in different namespaces

### ISSUE THREAD

**Issue by pshirshov** [circe/circe#868]
The following example fails:
```scala
import java.time._

import io.circe._
import io.circe.parser._
import io.circe.syntax._
import io.circe.generic.semiauto._
import io.circe.generic.decoding._
import io.circe.generic.encoding._
import io.circe.java8.time._

sealed trait Polymorphic

object NS1 {

  case class Payload(
                      zonedDateTime: ZonedDateTime = ZonedDateTime.now()
                      , utcZonedDateTime: ZonedDateTime = ZonedDateTime.now(ZoneId.of("UTC"))
                      , localDateTime: LocalDateTime = LocalDateTime.now()
                      , localTime: LocalTime = LocalTime.now()
                      , localDate: LocalDate = LocalDate.now()
                    ) extends Polymorphic

  object Payload {
    implicit val encodeTestPayload: Encoder[Payload] = deriveEncoder[Payload]
    implicit val decodeTestPayload: Decoder[Payload] = deriveDecoder[Payload]
  }

}

object NS2 {

  case class Payload(message: String) extends Polymorphic

  object Payload {
    implicit val encodeAnotherPayload: Encoder[Payload] = deriveEncoder[Payload]
    implicit val decodeAnotherPayload: Decoder[Payload] = deriveDecoder[Payload]
  }

}
object Polymorphic {
  implicit val encodePolymorphic: Encoder[Polymorphic] = deriveEncoder[Polymorphic]
  implicit val decodePolymorphic: Decoder[Polymorphic] = deriveDecoder[Polymorphic]
}

def test(t: Polymorphic): Unit = {
  val encoded = t.asJson.noSpaces
  println(s"Encoded:\n$encoded\n")
  val parsed = parse(encoded)
  println(s"Parsed:\n$parsed\n")
  val restored = parsed.map(_.as[Polymorphic])
  println(s"Restored:\n$restored\n")
}

test(NS1.Payload())
test(NS2.Payload("hi"))
```

While almost exactly the same one works:

```scala
import java.time._

import io.circe._
import io.circe.parser._
import io.circe.syntax._
import io.circe.generic.semiauto._
import io.circe.generic.decoding._
import io.circe.generic.encoding._
import io.circe.java8.time._

sealed trait Polymorphic

object NS1 {

  case class Payload(
                      zonedDateTime: ZonedDateTime = ZonedDateTime.now()
                      , utcZonedDateTime: ZonedDateTime = ZonedDateTime.now(ZoneId.of("UTC"))
                      , localDateTime: LocalDateTime = LocalDateTime.now()
                      , localTime: LocalTime = LocalTime.now()
                      , localDate: LocalDate = LocalDate.now()
                    ) extends Polymorphic

  object Payload {
    implicit val encodeTestPayload: Encoder[Payload] = deriveEncoder[Payload]
    implicit val decodeTestPayload: Decoder[Payload] = deriveDecoder[Payload]
  }

}

object NS2 {

  case class Payload1(message: String) extends Polymorphic

  object Payload1 {
    implicit val encodeAnotherPayload: Encoder[Payload1] = deriveEncoder[Payload1]
    implicit val decodeAnotherPayload: Decoder[Payload1] = deriveDecoder[Payload1]
  }

}
object Polymorphic {
  implicit val encodePolymorphic: Encoder[Polymorphic] = deriveEncoder[Polymorphic]
  implicit val decodePolymorphic: Decoder[Polymorphic] = deriveDecoder[Polymorphic]
}

def test(t: Polymorphic): Unit = {
  val encoded = t.asJson.noSpaces
  println(s"Encoded:\n$encoded\n")
  val parsed = parse(encoded)
  println(s"Parsed:\n$parsed\n")
  val restored = parsed.map(_.as[Polymorphic])
  println(s"Restored:\n$restored\n")
}

test(NS1.Payload())
test(NS2.Payload1("hi"))
```

As you may notice the only difference is the name of dummy class in `NS2`.

I understand why is it happening, though it's a bit weird and a bit inconvenient.

**Comment by travisbrown** [comment]
By "fails" you mean you're getting a compiler error for the first version? I think you're seeing the same issue as #855, which is due to an underlying issue in [Shapeless](https://github.com/circe/circe/issues/855#issuecomment-372502585). Ideally (I think—I'm not entirely sure what I want here, personally) Shapeless would disambiguate these (and possibly more generally include more information in the union keys for nested structures like this), and this code would work in circe entirely for free. Unfortunately as things are right now I don't have a good answer.

**Comment by pshirshov** [comment]
> By "fails" you mean you're getting a compiler error for the first version?
Exactly.

> Unfortunately as things are right now I don't have a good answer.
I understand. And yes, that's the same issue.

**Comment by travisbrown** [comment]
Closing as a duplicate of #855, which is blocked by Shapeless.
